src/service.rs

```diff
@@ -0,0 +1,327 @@
+    pub ports: HashMap<u16, u16>,
+    #[serde(default)]
+    pub endpoints: HashMap<NetworkAddress, Endpoint>,
+}
+
+impl Service {
+    pub fn namespaced_hostname(&self) -> NamespacedHostname {
+        NamespacedHostname {
+            namespace: self.namespace.clone(),
+            hostname: self.hostname.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Endpoint {
+    pub address: NetworkAddress,
+    pub port: HashMap<u16, u16>,
+}
+
+impl TryFrom<&XdsService> for Service {
+    type Error = WorkloadError;
+
+    fn try_from(s: &XdsService) -> Result<Self, Self::Error> {
+        let mut nw_addrs = Vec::new();
+        for addr in &s.addresses {
+            let network_address = network_addr(
+                &addr.network,
+                byte_to_ip(&Bytes::copy_from_slice(&addr.address))?,
+            );
+            nw_addrs.push(network_address);
+        }
+        let svc = Service {
+            name: s.name.to_string(),
+            namespace: s.namespace.to_string(),
+            hostname: s.hostname.to_string(),
+            addresses: nw_addrs,
+            ports: (&PortList {
+                ports: s.ports.clone(),
+            })
+                .into(),
+            endpoints: Default::default(), // intentionally empty; will be populated once inserted into the workload store
+        };
+        Ok(svc)
+    }
+}
+
+/// Data store for service information.
+#[derive(serde::Serialize, Default, Debug)]
+pub struct ServiceStore {
+    /// Maintains a mapping of service IP -> (workload IP -> workload endpoint)
+    /// this is used to handle ordering issues if workloads with VIPs are received before services.
+    staged_vips: HashMap<NetworkAddress, HashMap<NetworkAddress, Endpoint>>,
+
+    /// Maintains a mapping of workload IP to VIP. This is used only to handle removal of service
+    /// endpoints when a workload is removed.
+    workload_to_vips: HashMap<NetworkAddress, HashSet<NetworkAddress>>,
+
+    /// Allows for lookup of services by network address, the service's xds secondary key.
+    /// this map stores the same services as `services_by_hostname` so we can mutate the services in both maps
+    /// at once when a service is updated or workload updates affect a service's endpoints.
+    by_vip: HashMap<NetworkAddress, Arc<RwLock<Service>>>,
+
+    /// Allows for lookup of services by namespaced and hostname, the service's xds primary key.
+    /// this map stores the same services as `services_by_ip` so we can mutate the services in both maps
+    /// at once when a service is updated or workload updates affect a service's endpoints.
+    by_host: HashMap<String, Vec<Arc<RwLock<Service>>>>,
+}
+
+impl ServiceStore {
+    /// Returns the [Service] matching the given VIP.
+    pub fn get_by_vip(&self, vip: &NetworkAddress) -> Option<Arc<RwLock<Service>>> {
```

Comment:
not a huge fan of returning the `Arc<RwLock`'s themselves, since this is the underlying data store. this feels like a leaky abstraction. id prefer we return a copy of the data that the caller can do whatever it pleases with (despite the performance cost)

this also can make for sneaky deadlocks (e.g. two different callers of this function trying to lock the same lock, and it being unclear that they are using the same one)